### DISCUSSION TITLE
just-because-you-stopped-waiting-for-it-doesnt-mean-the-task-stopped-running/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#61] [Announcements]
# just-because-you-stopped-waiting-for-it-doesnt-mean-the-task-stopped-running/

In this post I look at what happens to Tasks that are 'cancelled' using the new .NET 6 WaitAsync() API, and in other Task cancellation scenarios

https://andrewlock.net/just-because-you-stopped-waiting-for-it-doesnt-mean-the-task-stopped-running/

**Comment by MatthewGabriel** [comment]
Thanks for this post Andrew.  Been looking into running a task with a timeout recently.  In order to enable the timeout to cancel the main task I'm building a helper method that runs the task and can therefore pass it a cancellation token that it controls in order to trigger cancellation.  Using a linked token etc.

I have a question about the final snippet of code you posted where you prevent the leak by cancelling the delay task.  My understanding is that cancelling a delay task would raise an exception.  So is it not possible that cancelling the delay and then walking away could trigger an unobservedtaskexception at some point later on?

My attempts to prove this have failed so far as what happens is very subtle depending on whether tasks are running within a Task.Run or not.  Perhaps you have a better understanding of tasks under the hood than I do.

Would it be safer to cancel the delay task, then await it to capture the exception, and then finally await the main task?

[a comment by an automated account is not shown]
